### ISSUE TITLE
[Suggestion] Removal of Client Color Configuration and finally Translate OneProbe

### ISSUE THREAD

**Issue by Speiger** [McJty/TheOneProbe#514]
Header is truth and Click-bait at the same time.

Seriously get rid of the Client Config in One Probe and breaking MC Text Rendering support.
Now this would be denied without a Solution... 

My solution is fairly simple. Instead of Injecting the Color in the client when it is received how about having it injected server sided.
And Player Choice isn't reduced by it.

On login send a Packet from the Player to the server that just has the Color Preferences included and when the config is updated update it on the server too.
When the Probe Data is requested you have a PerPlayerCache what colors are to use etc and that way you can use MCs Text system and still have Color Support. Because MC supports Color Rendering with RGB if you use a HTML String. (it converts HTML Color into RGB supported color in the Style.class)

Allowing for the same features but Properly Implemented instead of a Regex-hack to achieve the same thing.

Now this would be technically gamebreaking but only if you remove the client options fully instead of making it legacy.

Also this would allow translating One Probe.

**Comment by McJty** [comment]
Hmm, most of client config is pretty much required (like how long to wait for server timeout and so on).

What exactly are you thinking about? Which client side config options should be removed for you?

**Comment by Speiger** [comment]
Client Customization should be moved to the server Stuff like Text Coloring should be server side (per player) instead of Using hacks to get it into the client.
That way you can actually use the full textrendering and translation which is right now impossible to do so with the current client injection system

**Comment by McJty** [comment]
I asked for a concrete example of a client side config where text or text coloring is being set client side

**Comment by Speiger** [comment]
https://github.com/McJtyMods/TheOneProbe/blob/fdc11fac9b83b5bb62e6d61e450e5afdb222b2ba/src/main/java/mcjty/theoneprobe/apiimpl/client/ElementTextRender.java#L36
This function should be no longer used by OneProbe itself and just be legacy by default so other mods dont break.
This is the "Client Color Customization" i meant

**Comment by Speiger** [comment]
Right now you are basically breaking down TextComponents into raw strings to inject ColorCodes so the text is customize able by each player. That way you basically remove every single feature that MC supports because you delete all the ITextComponent things and just get the raw string instead. Very Limiting for a feature that could be replaced by just injecting this at server side.

**Comment by McJty** [comment]
Should be fairly easy to do this without breaking. If the text doesn't contain any STARTLOC/ENDLOC/{= or =} then we use directly render the ITextComponent. Otherwise we revert to old system. No breackage needed

**Comment by Speiger** [comment]
Yeah you just need to have a Login/ConfigChanged Packet (can be the same) and send the preferences over to the server.
And when the server constructs the ProbeData it just asks the server cache to get the colors instead of using CompoundText.
Just make sure its a PerPlayerCache.

Also maybe write it so that other mods can easily add their own customization templates in there like OneProbe.
Maybe keep track of it Via ResourceLocation so mods can just request their own template.

**Comment by McJty** [comment]
What? I'm confused about that. Why would there be the need for a packet? The text and style that is given to that stylifyString() doesn't come from a client. It comes from the server. i.e. it's whatever the mod using the TOP API puts in the string in its info provider

**Comment by Speiger** [comment]
The System as it is right now:
You build a CompoundText and put prefixes in then you send this compound text (as ITextComponent) to the client
then you extract the string out of the ITextComponent LOSING "ALL" Style information because that is no longer inside of the getString method and then you inject all the Overrides (ClientSided) in your string before rendering it inside a legacy method.

What i suggest is: You send the Color Customization to the server and when you build a text you don't use the CompoundText but instead you use ITextComponent and put the TextColor in there using Style.setColor (That supports RGB) and send that to the client and have full ITextComponent support... No Loss of data required.

Translation, Coloring, Custom Styles, etc.

**Comment by McJty** [comment]
yes. That usage of prefixes should be deprecated. But mods can already use the new ITextComponent API which contains styling. So they should switch to that and we just make sure that we don't throw away that info unless they use the old deprecated STARTLOC and other styling. But that's just an internal change to TOP. There is no need for an extra packet to send over stuff from client to server. I wouldn't know what to send there

**Comment by McJty** [comment]
Ah I finally found what you were refering too (which was the example I asked for in the beginning). You're talking about the [style] section in theoneprobe-client.toml. I missed that one

I have to think about that one though. Not high priority at the moment either

**Comment by Speiger** [comment]
Would you accept pullrequests in 1.16 for this? (You can port it i am stuck in 1.16 for now)

**Comment by McJty** [comment]
If they are done well and compatible

**Comment by McJty** [comment]
TextStyleClass is a consideration here. The idea here is that server side the mods use logical styles like NAME, INFO, WARNING, ERROR, ... and client side that is configered to be a certain style. That functionality should remain in some way

**Comment by Speiger** [comment]
Once i am back to mc modding (right now trying to reach a different milestone in a different project)
I can try look into that one. (Unless you beat me to it).

Also I will most likely also translate OneProbe with it because language support should have been a thing day one.

**Comment by McJty** [comment]
Well the reason I mention is because you can't separate that from what you're trying to do. It sits within the same routines
